File: leetcode/p1631.cpp
class Solution {
public:
    priority_queue<pair<int, pair<int, int>>> pq;
    void push(vector<vector<int>>& heights, int x, int y, int h, int max_df) {
        int df = abs(h - heights[x][y]);
        max_df = max(max_df, df);
        //cout << x << y << df << max_df << endl;
        pq.push({-1*(max_df), {x, y}});
    }
    int minimumEffortPath(vector<vector<int>>& heights) {
        int row=heights.size(), col=heights[0].size();
        vector<vector<bool>> visited(row, vector<bool>(col, false));
        pq.push({0, {row-1, col-1}});
        while (!pq.empty()) {
            auto [v, p] = pq.top();
            pq.pop();
            //cout << p.first << p.second << v << endl;
            if (visited[p.first][p.second]) continue;
            if (p.first == 0 && p.second == 0) return -v;
            int h = heights[p.first][p.second];
            visited[p.first][p.second] = true;
            if (p.first+1<row) push(heights, p.first+1, p.second, h, -v);
            if (p.second+1<col) push(heights, p.first, p.second+1, h, -v);
            if (p.first-1>=0) push(heights, p.first-1, p.second, h, -v);
            if (p.second-1>=0) push(heights, p.first, p.second-1, h, -v);
        }
        return -1;
    }
};

File: leetcode/p1631.py
class Solution:
    def check_update(self, x0, y0, x1, y1):
        df = abs(self.heights[x0][y0] - self.heights[x1][y1])
        df = max(self.table[x0][y0], df)
        if df < self.table[x1][y1]:
            self.table[x1][y1] = df
            heapq.heappush(self.qheap, (df, x1, y1))
    
    def run_update(self, x, y):
        if x > 0:
            self.check_update(x, y, x-1, y)
        if y > 0:
            self.check_update(x, y, x, y-1)
        if x < len(self.heights) - 1:
            self.check_update(x, y, x+1, y)
        if y < len(self.heights[0]) - 1:
            self.check_update(x, y, x, y+1)
                
    def minimumEffortPath(self, heights: List[List[int]]) -> int:
        self.heights = heights
        self.table = [[float('inf')]*len(heights[0]) for i in range(len(heights))]
        self.table[len(heights)-1][len(heights[0])-1] = 0
        self.qheap = [(0, len(heights)-1, len(heights[0])-1)]
        self.visited = set()
        while self.qheap:
            _, x, y = heapq.heappop(self.qheap)
            if (x, y) in self.visited:
                continue
            self.visited.add((x, y))
            self.run_update(x, y)
        #print(self.table)
        return self.table[0][0]
